<a href="https://colab.research.google.com/github/I-Love-Ducks/Data-science/blob/main/Arhaan_Pandas_Worksheet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Pandas Part 1

In the NumPy section we dealt with some arrays, whose columns had each a special meaning. For example, the column number 0 could contain values interpreted as years, and column 1 could contain a month, and so on. It is possible to handle the data this way, but in can be **hard to remember**, which column number corresponds to which variable. Especially, if you later **remove some column from the array, then the numbering of the remaining columns changes**. One solution to this is to give a **descriptive name to each column**. **These column names stay fixed and attached to their corresponding columns, even if we remove some of the columns**. In addition, the **rows** can be given **names** as well, these are called **indices** in Pandas.

The Pandas library is built on top of the NumPy library, and it provides a special kind of two dimensional data structure called DataFrame. **The DataFrame allows to give names to the columns, so that one can access a column using its name in place of the index of the column.**

In [2]:
pip install pandas

In [3]:
import pandas as pd    # This is the standard way of importing the Pandas library
import numpy as np

## Basics of DataFrame

Let's import some weather data that is in text form in a csv (Comma Separated Values) file. The following call will fetch the data from the internet and convert it to a DataFrame:

In [4]:
wh = pd.read_csv("https://raw.githubusercontent.com/csmastersUH/data_analysis_with_python_2020/master/kumpula-weather-2017.csv")
wh.head()   # The head method prints the first 5 rows

Year  m  d   Time Time zone  Precipitation amount (mm)  Snow depth (cm)  \
0  2017  1  1  00:00       UTC                       -1.0             -1.0   
1  2017  1  2  00:00       UTC                        4.4             -1.0   
2  2017  1  3  00:00       UTC                        6.6              7.0   
3  2017  1  4  00:00       UTC                       -1.0             13.0   
4  2017  1  5  00:00       UTC                       -1.0             10.0   

   Air temperature (degC)  
0                     0.6  
1                    -3.9  
2                    -6.5  
3                   -12.8  
4                   -17.8

To get a single column, we refer to it by its name:

In [5]:
wh["Snow depth (cm)"].head()

0    -1.0
1    -1.0
2     7.0
3    13.0
4    10.0
Name: Snow depth (cm), dtype: float64

We can also get the mean of a column:

In [6]:
wh["Air temperature (degC)"].mean()

6.527123287671233

To remove a column, we can use the .drop function:

In [7]:
wh.drop("Time zone", axis=1).head() # returns a copy without the Time zone column
# Remember to use axis = 1, because otherwise it will look for a row named "Time Zone"

Year  m  d   Time  Precipitation amount (mm)  Snow depth (cm)  \
0  2017  1  1  00:00                       -1.0             -1.0   
1  2017  1  2  00:00                        4.4             -1.0   
2  2017  1  3  00:00                        6.6              7.0   
3  2017  1  4  00:00                       -1.0             13.0   
4  2017  1  5  00:00                       -1.0             10.0   

   Air temperature (degC)  
0                     0.6  
1                    -3.9  
2                    -6.5  
3                   -12.8  
4                   -17.8

In [8]:
# notice how above, I said that it returns a copy. The original is unchanged.
wh.head()

Year  m  d   Time Time zone  Precipitation amount (mm)  Snow depth (cm)  \
0  2017  1  1  00:00       UTC                       -1.0             -1.0   
1  2017  1  2  00:00       UTC                        4.4             -1.0   
2  2017  1  3  00:00       UTC                        6.6              7.0   
3  2017  1  4  00:00       UTC                       -1.0             13.0   
4  2017  1  5  00:00       UTC                       -1.0             10.0   

   Air temperature (degC)  
0                     0.6  
1                    -3.9  
2                    -6.5  
3                   -12.8  
4                   -17.8

To add a new column, you make one like you would in a python dictionary:

In [9]:
wh["Rainy"] = wh["Precipitation amount (mm)"] > 5
wh.head()

Year  m  d   Time Time zone  Precipitation amount (mm)  Snow depth (cm)  \
0  2017  1  1  00:00       UTC                       -1.0             -1.0   
1  2017  1  2  00:00       UTC                        4.4             -1.0   
2  2017  1  3  00:00       UTC                        6.6              7.0   
3  2017  1  4  00:00       UTC                       -1.0             13.0   
4  2017  1  5  00:00       UTC                       -1.0             10.0   

   Air temperature (degC)  Rainy  
0                     0.6  False  
1                    -3.9  False  
2                    -6.5   True  
3                   -12.8  False  
4                   -17.8  False

We will get more to this in the second part of Pandas, but this should be the very basics.

---



## Series

A series is a Dataframe, but it is one-dimensional.

In [10]:
s = pd.Series([1,3,2,5,6])
s
# The left numbers show the index, and the right side shows the values

0    1
1    3
2    2
3    5
4    6
dtype: int64

We can also add a name to our series:

In [11]:
s.name = "Random numbers"
s

0    1
1    3
2    2
3    5
4    6
Name: Random numbers, dtype: int64

These are the most common attributs of a series:

In [12]:
print(f"Name: {s.name}, dtype: {s.dtype}, size: {s.size}")

Name: Random numbers, dtype: int64, size: 5


We can also index and slice our series:

In [13]:
s[0], s[1]

(1, 3)

In [14]:
s2 = s[2:] # from index 2 to end
s2

2    2
3    5
4    6
Name: Random numbers, dtype: int64

In [15]:
s[-1:] # back one row till the end

4    6
Name: Random numbers, dtype: int64

We can show where the indices start and stop with step from the original series:

In [16]:
s2.index

RangeIndex(start=2, stop=5, step=1)

We can also change the indexing:

In [17]:
s3=pd.Series([1, 4, 5, 2, 5, 2], index=list("abcdef"))
s3.name = "Letter Indexes"
s3

a    1
b    4
c    5
d    2
e    5
f    2
Name: Letter Indexes, dtype: int64

We can also index and slice based on the indexes we make:

In [18]:
s3["b"]

4

In [19]:
s3["d":"f"]

d    2
e    5
f    2
Name: Letter Indexes, dtype: int64

You can also concatenate Series' together using the pd.concat:

In [20]:
s4 = pd.concat([s3, pd.Series([4,3,2], index = ["g","h","i"]) ])
s4

a    1
b    4
c    5
d    2
e    5
f    2
g    4
h    3
i    2
dtype: int64

or just add a new index like a dictionary:

In [21]:
s3["L"] = 13
s3

a     1
b     4
c     5
d     2
e     5
f     2
L    13
Name: Letter Indexes, dtype: int64

To delete an element, use the del keyword

In [22]:
del s3["L"]
s3

a    1
b    4
c    5
d    2
e    5
f    2
Name: Letter Indexes, dtype: int64

# Part one Exercises

##Read series

Write function read_series that reads input lines from the user and return a Series. Each line should contain first the index and then the corresponding value, separated by whitespace. The index and values are strings (in this case dtype is object). An empty line signals the end of Series. Malformed input should cause an exception. An input line is malformed, if it is non-empty and, when split at whitespace, does not result in two parts.

Test your function:

In [23]:
#def read_series():

In [24]:
#read_series()

This was my answer to the problem:

<font color = "white">def read_series():

<font color = "white">  x = input("Please provide a list of strings seperated by spaces: ")

<font color = "white">  organize = list(x.split(sep = " "))

<font color = "white">  print(pd.Series(organize))</font>

##Operations on series

Write function create_series that gets two lists of numbers as parameters. Both lists should have length 3. The function should first create two Series, s1 and s2. The first series should have values from the first parameter list and have corresponding indices a, b, and c. The second series should get its values from the second parameter list and have again the corresponding indices a, b, and c. The function should return the pair of these Series.

Then, write a function modify_series that gets two Series as parameters. It should add to the first Series s1 a new value with index d. The new value should be the same as the value in Series s2 with index b. Then delete the element from s2 that has index b. Now the first Series should have four values, while the second list has only two values. Adding a new element to a Series can be achieved by assignment, like with dictionaries. Deletion of an element from a Series can be done with the del statement.

Test these functions from the main function. Try adding together the Series returned by the modify_series function. The operations on Series use the indices to keep the element-wise operations aligned. If for some index the operation could not be performed, the resulting value will be NaN (Not A Number).

In [25]:
#create_series():

In [26]:
#modify_series():

In [27]:
#Try them out!

###Answer

In [28]:
def create_series(alist, blist):
  a = list(alist.split(sep = " "))
  b = list(blist.split(sep = " "))
  s1 = pd.Series(a, index = ["a", "b", "c"])
  s2 = pd.Series(b, index = ["a", "b", "c"])
  return s1,s2

In [29]:
def modify_series(aser, bser):
  aser["d"] = bser["b"]
  del bser["b"]

In [30]:
#list1 = input("List 1(Seperated by spaces): ")
#list2 = input("List 2(Seperated by spaces): ")
#x, y = create_series(list1, list2)
#modify_series(x,y)
#x,y


## Inverse series

Write function inverse_series that get a Series as a parameter and returns a new series, whose indices and values have swapped roles. Test your function from the main function.

In [31]:
#def inverse_series():

In [32]:
#Try it out!

###Answer

In [33]:
def inverse_series(ser):
  key = ser.keys()
  val = ser.values
  return pd.Series(key, index = val)

In [34]:
d = pd.Series([1, 2, 3], index = ["a", "b", "c"])
inverse_series(d)

1    a
2    b
3    c
dtype: object

#Pandas Part 2

##DataFrames(again)

To make a dataframe, we use the pd.DataFrame function:

In [35]:
df = pd.DataFrame(np.random.randn(2,3), columns = ["First", "Second", "Third"], index = ["a","b"]) # dont worry about randn; randn is just a way to make random numbers
df

First    Second     Third
a  1.674898  1.881299 -0.144933
b -0.501754 -1.303765  1.022447

As seen above, the columns argument changes the names of the columns:

In [36]:
df.columns

Index(['First', 'Second', 'Third'], dtype='object')

And the index argument is the names of the rows:

In [37]:
df.index

Index(['a', 'b'], dtype='object')

If we leave out the columns argument, The columns' name will be defaulted to a list of integers:

In [38]:
df2 = pd.DataFrame(np.random.randn(2,3), index = ["a", "b"])
df2

0        1         2
a -1.682014  0.00528 -0.228310
b -0.150213 -0.93757 -0.591273

**To make a dataframe from columns, do this:**

In [39]:
s1 = pd.Series([1,2,3])
s1

0    1
1    2
2    3
dtype: int64

In [40]:
s2 = pd.Series([4,5,6], name="b")
s2

0    4
1    5
2    6
Name: b, dtype: int64

In [41]:
pd.DataFrame(s1, columns=["a"])

a
0  1
1  2
2  3

As you can see, the indexes of the series became the rows, and the column name was specified.
\
\
Below is an example without specifying the columns:

In [42]:
pd.DataFrame(s2)

b
0  4
1  5
2  6

This works because we named s2 "b" and so the name got reflected on the column.

**To make a dataframe from rows, do this:**

If the rows are {"Jake", 12, 9} and {"Jack", 14, 5}, with columns "Name", "Age", "Work Experience", do this:

In [43]:
pd.DataFrame([{"Name":"Jake", "Age":"12", "WA":"9"}, {"Name":"Jack", "Age":"14", "WA":"5"}])

Name Age WA
0  Jake  12  9
1  Jack  14  5

Or this:

In [44]:
df = pd.DataFrame([[1000, "Jack", 21], [1500, "John", 29]], columns=["Wage", "Name", "Age"])
df

Wage  Name  Age
0  1000  Jack   21
1  1500  John   29

###Accessing Rows or Columns in a DataFrame

Using lists, to access a row, you would just use [] and the number index.

To access rows or columns in DataFrames, it's little weird.

**Columns:**

If you want to use a integer, like ```df[1]```, it will access the column, but only if the name is explicitly the integer.

Ex:

In [45]:
np.random.seed(1)
df = pd.DataFrame(np.random.randn(2,3), columns = [1, 2, 3])
df[1] # this works because there is a column named 1

0    1.624345
1   -1.072969
Name: 1, dtype: float64

The code below will display an error

In [46]:
#df = pd.DataFrame(np.random.randn(2,3), columns = ["a", "b", "c"])
#df[1] # this does not work because there is no column named 1

To index the code above, we can use the name of the column:

In [47]:
df = pd.DataFrame(np.random.randn(2,3), columns = ["a", "b", "c"])
df["a"] # this works as we used "a", which is the column name

0    1.744812
1   -0.249370
Name: a, dtype: float64

**Rows:**

If we index a slice or a boolean mask, then the index will apply to the rows

In [48]:
df

a         b         c
0  1.744812 -0.761207  0.319039
1 -0.249370  1.462108 -2.060141

In [49]:
df[0:1] # index from row 0 to 1. (1 not incl.)

a         b         c
0  1.744812 -0.761207  0.319039

In [50]:
df[df.a > 1] # the row in which "a" is greater that 1

a         b         c
0  1.744812 -0.761207  0.319039

We can also index through multiple dimensions:

In [51]:
df["a"][1] # note the order of dimensions

-0.2493703754774101

But instead of the above code, we can use the .loc[] and the .iloc[] functions.

In [52]:
df.loc[1, "a"] # use square brackets and be sure to use the row before the column

-0.2493703754774101

The difference between loc and iloc is that loc can use numbers and strings, like the ```"a"``` in the above code, but iloc can only use numbers.

In [53]:
df.iloc[1, 0] # "a" was the first column, so I used 0 to imply that

-0.2493703754774101

In [54]:
df.iloc[-1,-1]

-2.060140709497654

In [55]:
df.loc[1, ["a", "b"]] # You can show multiple rows and columns

a   -0.249370
b    1.462108
Name: 1, dtype: float64

In [56]:
df.loc[[0,1], ["a", "b", "c"]]

a         b         c
0  1.744812 -0.761207  0.319039
1 -0.249370  1.462108 -2.060141

### Statistics and MIssing data

In [57]:
wh = pd.read_csv("https://raw.githubusercontent.com/csmastersUH/data_analysis_with_python_2020/master/kumpula-weather-2017.csv")
wh.head()

Year  m  d   Time Time zone  Precipitation amount (mm)  Snow depth (cm)  \
0  2017  1  1  00:00       UTC                       -1.0             -1.0   
1  2017  1  2  00:00       UTC                        4.4             -1.0   
2  2017  1  3  00:00       UTC                        6.6              7.0   
3  2017  1  4  00:00       UTC                       -1.0             13.0   
4  2017  1  5  00:00       UTC                       -1.0             10.0   

   Air temperature (degC)  
0                     0.6  
1                    -3.9  
2                    -6.5  
3                   -12.8  
4                   -17.8

Finding the mean of everything is time-consuming and ineffiecient:

In [58]:
wh2 = wh.drop(["Year", "m", "d"], axis=1) # remove year, month, and day
snowAvg = wh2["Snow depth (cm)"].mean()
precip = wh2["Precipitation amount (mm)"].mean()
airTemp = wh2["Air temperature (degC)"].mean()

print("Snow Depth:", snowAvg)
print("Precipitation amount:", precip)
print("Air temperature:", airTemp)

Snow Depth: 0.9664804469273743
Precipitation amount: 1.9663013698630138
Air temperature: 6.527123287671233


The above code is super boring to write. Instead we can use the .describe() function. This shows all the stats you need for all the columns:

In [59]:
wh.describe()

Year           m           d  Precipitation amount (mm)  \
count   365.0  365.000000  365.000000                 365.000000   
mean   2017.0    6.526027   15.720548                   1.966301   
std       0.0    3.452584    8.808321                   4.858423   
min    2017.0    1.000000    1.000000                  -1.000000   
25%    2017.0    4.000000    8.000000                  -1.000000   
50%    2017.0    7.000000   16.000000                   0.200000   
75%    2017.0   10.000000   23.000000                   2.700000   
max    2017.0   12.000000   31.000000                  35.000000   

       Snow depth (cm)  Air temperature (degC)  
count       358.000000              365.000000  
mean          0.966480                6.527123  
std           3.717472                7.183934  
min          -1.000000              -17.800000  
25%          -1.000000                1.200000  
50%          -1.000000                4.800000  
75%           0.000000               12.900000  
max          15.000000               19.600000

The -1 means 0, and the 0 means close to zero.

Now, you might have noticed something wrong. In the count column, showing how many days it was recorded, the snow depth is only 358, while all the others are 365. Why?

In [60]:
wh["Snow depth (cm)"].unique() # .unique shows all the values without any repitition of values

array([-1.,  7., 13., 10., 12.,  9.,  8.,  5.,  6.,  4.,  3., 15., 14.,
        2., nan,  0.])

Look at that. NaN. This means Not a Number. This can happen when there is a impossible calculation, like 0/0, or if the data is actually not a number.

NaN can only happen in a float Series or DataFrame. If we make an integer Series, but add a NaN, it gets promoted to float:

In [61]:
ser = pd.Series([1, 2, 3])
ser # look at the dtype

0    1
1    2
2    3
dtype: int64

In [62]:
ser = pd.Series([1, 2, 3, np.nan])
ser # look at the dtype

0    1.0
1    2.0
2    3.0
3    NaN
dtype: float64

Lets say we want to see where are the missing values. We can use the .isnull() function to help us

In [63]:
wh.isnull()

Year      m      d   Time  Time zone  Precipitation amount (mm)  \
0    False  False  False  False      False                      False   
1    False  False  False  False      False                      False   
2    False  False  False  False      False                      False   
3    False  False  False  False      False                      False   
4    False  False  False  False      False                      False   
..     ...    ...    ...    ...        ...                        ...   
360  False  False  False  False      False                      False   
361  False  False  False  False      False                      False   
362  False  False  False  False      False                      False   
363  False  False  False  False      False                      False   
364  False  False  False  False      False                      False   

     Snow depth (cm)  Air temperature (degC)  
0              False                   False  
1              False                   False  
2              False                   False  
3              False                   False  
4              False                   False  
..               ...                     ...  
360            False                   False  
361            False                   False  
362            False                   False  
363            False                   False  
364            False                   False  

[365 rows x 8 columns]

Obviously this is not effeicient because of the quantity of rows, so the below code can help:

In [64]:
wh[wh.isnull().any(axis=1)] # finds values with NaN and displays them

Year   m   d   Time Time zone  Precipitation amount (mm)  \
74   2017   3  16  00:00       UTC                        1.8   
163  2017   6  13  00:00       UTC                        0.6   
308  2017  11   5  00:00       UTC                        0.2   
309  2017  11   6  00:00       UTC                        2.0   
313  2017  11  10  00:00       UTC                        3.6   
321  2017  11  18  00:00       UTC                       11.3   
328  2017  11  25  00:00       UTC                        8.5   

     Snow depth (cm)  Air temperature (degC)  
74               NaN                     3.4  
163              NaN                    12.6  
308              NaN                     8.4  
309              NaN                     7.5  
313              NaN                     7.2  
321              NaN                     5.9  
328              NaN                     4.2

The dropna function is a function that drops columns or rows **that contain missing values**, based on the axis argument:

In [65]:
wh.dropna().shape # Default axis is 0(rows)

(358, 8)

In [66]:
wh.dropna(axis=1).shape # removes the columns with a NaN, which is only snow depth

(365, 7)

The how and thresh parameters of the dropna method allow one to specify how many values need to be missing in order for the row/column to be dropped. **I'm not going to cover this.**

The problem is, what do we do with these NaNs without manually changing each of them? The ```.fillna()``` function will help us.

The fillna function has two fill types:

```df = df.fillna(method = 'ffill')```:

This fills any NaNs with the value right before it \

```df = df.fillna(method = 'bfill')```:

This will fill any NaN with the value right after it \

In [67]:
wh = wh.fillna(method='ffill')
wh[wh.isnull().any(axis=1)] # doesn't detect any NaNs after filling

<ipython-input-67-901454edcd83>:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  wh = wh.fillna(method='ffill')


Empty DataFrame
Columns: [Year, m, d, Time, Time zone, Precipitation amount (mm), Snow depth (cm), Air temperature (degC)]
Index: []

There is another way of filling NaNs called interpolation, but I can't cover that right now, so here is the documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html

### Converting columns

When we want to change the type of a column or multiple columns, we must use the .map() or .astype() methods.

The .map() method is only used when converting one column(series).

The .astype() method is used to convert many columns(dataframe).

Some examples:

In [68]:
ser = pd.Series(["1","2"])
print(ser) # notice the type
print("\n\n")
ser = ser.map(int)
print(ser) # notice the type

0    1
1    2
dtype: object



0    1
1    2
dtype: int64


In [69]:
df = pd.DataFrame([[1,2,3],[4,5,6],[7,8,9]], columns = ["a","b","c"])
df

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

In [70]:
dfs = df.astype(str)
print(dfs)
dfs.dtypes

   a  b  c
0  1  2  3
1  4  5  6
2  7  8  9


a    object
b    object
c    object
dtype: object

The .to_numeric() can change the values to integer or float based on the downcast

In [71]:
pd.to_numeric(pd.Series([1,1.0]), downcast="integer")

0    1
1    1
dtype: int8

In [72]:
pd.to_numeric(pd.Series([1,"a"]), errors="coerce")      # conversion error produces NaN

0    1.0
1    NaN
dtype: float64

In [73]:
df = pd.DataFrame([[1,2,3],[4,5,6],[7,8,9]], columns = ["a","b","c"])
df

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

In [74]:
df = df.astype({"b" : float, "c" : str})    # different types for columns
print(df.dtypes)
print(df)

a      int64
b    float64
c     object
dtype: object
   a    b  c
0  1  2.0  3
1  4  5.0  6
2  7  8.0  9


### String Processes

The .str attribute of a Series can help process the string differently. For example, the .str.capitalize() method:

In [75]:
names = pd.Series(["donald", "theresa", "angela", "vladimir"])
names.str.capitalize()

0      Donald
1     Theresa
2      Angela
3    Vladimir
dtype: object

To split the Series into different elements, use the .split command:

In [76]:
full_names = pd.Series(["Donald Trump", "Theresa May", "Angela Merkel", "Vladimir Putin"])
full_names.str.split()

0      [Donald, Trump]
1       [Theresa, May]
2     [Angela, Merkel]
3    [Vladimir, Putin]
dtype: object

The expand argument makes the elements split into different columns:

In [79]:
full_names.str.split(expand=True)

0       1
0    Donald   Trump
1   Theresa     May
2    Angela  Merkel
3  Vladimir   Putin

#Part two exercises

##Cities

Write function cities that returns the following DataFrame of top Finnish cities by population:

Should print:



```
                 Population Total area
Helsinki         643272     715.48
Espoo            279044     528.03
Tampere          231853     689.59
Vantaa           223027     240.35
Oulu             201810     3817.52
```



Answer(Select to see):


<font color = "white">hels = [643272, 715.48]

<font color = "white">esp = [279044, 528.03]

<font color = "white">tamp = [231853, 689.59]

<font color = "white">vant = [223027, 240.35]

<font color = "white">oulu = [201810, 3817.52]

<font color = "white">pd.DataFrame([hels, esp, tamp, vant, oulu], columns=["Population", "Total area"], index = ["Helsinki", "Espoo", "Tampere", "Vantaa", "Oulu"])</font>

##Powers of series

Make function ```powers_of_series``` that takes a Series and a positive integer k as parameters and returns a DataFrame. The resulting DataFrame should have the same index as the input Series. The first column of the dataFrame should be the input Series, the second column should contain the Series raised to power of two. The third column should contain the Series raised to the power of three, and so on until (and including) power of k. The columns should have indices from 1 to k.

The values should be numbers, but the index can have any type. Test your function from the main function. Example of usage:


```
s = pd.Series([1,2,3,4], index=list("abcd"))
print(powers_of_series(s, 3))
```

Would output:

```
   1   2   3
a  1   1   1
b  2   4   8
c  3   9  27
d  4  16  64
```


In [77]:
#def powers_of_series(args):

In [78]:
#s = pd.Series([1,2,3,4], index=list("abcd"))
#print(powers_of_series(s, 3))

Answer(Select to see):
<font color = "white">

<font color = "white">def powers_of_series(ser, col):
  
  <font color = "white">temp = pd.DataFrame(ser)
  
  <font color = "white">for i in range(col):
    
  <font color = "white">temp[i+1] = temp[0]**(i+1) # making a new column by taking the first one and exponentiated it


  <font color = "white">del temp[0] # removing extra column

  
 <font color = "white"> return temp

#Sources Cited:

Mooc.fi, 2024, courses.mooc.fi/org/uh-cs/courses/dap-22/chapter-4/pandas-part-2. Accessed 28 Sept. 2024.
